File: JavaScript/p0003.js
/**
 * @param {string} s
 * @return {number}
 */

const lengthOfLongestSubstring = s => {
  let answer = 0;
  let count = 0;
  let hash = {};

  if (s.length < 2) {
    return s.length;
  }

  for (let i = 0; i < s.length; i++) {
    if (hash[s[i]] == null) {
      count += 1;
    } else {
      count = Math.min(i - hash[s[i]], count + 1);
    }
    answer = Math.max(answer, count);
    hash[s[i]] = i;
  }
  return answer;
};

File: Go/p0003.go
package leetcode

import (
	"math"
)

func lengthOfLongestSubstring(s string) int {
	rs := []rune(s)
	if len(rs) < 2 {
		return len(rs)
	}
	answer := 0
	count := 0
	m := make(map[rune]int)
	for i, r := range rs {
		if _, ok := m[r]; !ok {
			count++
		} else {
			count = int(math.Min(float64(i-m[r]), float64(count+1)))
		}
		answer = int(math.Max(float64(answer), float64(count)))
		m[r] = i
	}
	return answer
}
